src/core/drive/form_submission.ts

```diff
@@ -65,61 +56,27 @@ export class FormSubmission {
     return Promise.resolve(confirm(message))
   }
 
-  constructor(
-    delegate: FormSubmissionDelegate,
-    formElement: HTMLFormElement,
-    submitter?: HTMLElement,
-    mustRedirect = false
-  ) {
+  constructor(delegate: FormSubmissionDelegate, submission: HTMLFormSubmission, mustRedirect = false) {
     this.delegate = delegate
-    this.formElement = formElement
-    this.submitter = submitter
-    this.formData = buildFormData(formElement, submitter)
-    this.location = expandURL(this.action)
-    if (this.method == FetchMethod.get) {
-      mergeFormDataEntries(this.location, [...this.body.entries()])
-    }
+    this.submission = submission
+    this.formElement = submission.form
+    this.submitter = submission.submitter
+    this.formData = submission.formData
+    this.location = submission.location
+    this.method = submission.fetchMethod
```

Comment:
The majority of these `readonly` properties were once defined as `get`-powered properties, which are part of the `FormSubmission` public API. Since `turbo:submit-start` and `turbo:submit-end` event listeners are able to access `FormSubmission` instances, it feels like we should go through a deprecation cycle before moving them to be `private` or replaced by `this.submission[...]` lookups.

Reply:
If the next version of Turbo is a major version bump to `8.0`, that might have implications for deprecation.